infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -497,6 +436,79 @@ private void clearStreams(Set<UUID> streamsToClear) {
         }
     }
 
+    private void batchingData(List<SMREntry> smrEntries, long currentSeqNum, UUID streamId, UUID snapshotSyncId,
+                              boolean applyPhase) {
+        CorfuStoreMetadata.Timestamp ts = null;
+        List<SMREntry> buffer = new ArrayList<>();
+        long bufferSize = 0;
+        int numBatches = 1;
+        int metadataSize = applyPhase ? 0 : metadataManager.getReplicationMetadata(session).getSerializedSize();
+
+        for (SMREntry smrEntry : smrEntries) {
+            // Apply all SMR entries in a single transaction as long as it does not exceed maxApplySize(a fraction of
+            // the runtime's max uncompressed write size(100 MB by default).  The fraction is added as a buffer to
+            // account for extra bytes added during LogData.serialize()).
+            // It was observed that special streams(ProtobufDescriptor table), can get a lot of updates, especially
+            // due to schema updates during an upgrade.  If the table was not checkpointed and trimmed on the Source,
+            // no de-duplication on these updates will occur.  As a result, the transaction size can be large.
+            // Although it is within the maxApplySize limit, deserializing these entries to read the table can cause an
+            // OOM on applications running with a small memory footprint.  So for such tables, introduce an
+            // additional limit of max number of entries(50 by default) applied in a single transaction.  This
+            // algorithm is in line with the limits imposed in Compaction and Restore workflows.
+            if (bufferSize + smrEntry.getSerializedSize() + metadataSize > replicationContext.getConfig(session).getMaxApplySize()
+                    || maxEntriesLimitReached(streamId,
+                    buffer)) {
+                try (TxnContext txnContext = metadataManager.getTxnContext()) {
+                    updateLog(txnContext, buffer, streamId);
+                    if (!applyPhase) {
```

Comment:
Looks like we can abstract out the metadata update for transfer phase since it is used twice